internal/tools/changes/repository.go

```diff
@@ -21,42 +32,249 @@ func NewRepository(path string) (*Repository, error) {
 	repo := &Repository{
 		RootPath: path,
 		Metadata: metadata,
+		git: git.Client{
+			RepoPath: path,
+		},
+		golist: golist.Client{
+			RootPath:       path,
+			ShortenModPath: shortenModPath,
+		},
+		Logf: log.Printf,
 	}
 
 	return repo, nil
 }
 
+// Modules returns all Go modules under the given Repository. The list of module names returns are shortened to be relative
+// to the root of the repository.
+func (r *Repository) Modules() ([]string, error) {
+	if r.modules != nil {
+		return r.modules, nil
+	}
+
+	mods, _, err := discoverModules(r.golist, r.RootPath)
+	if err != nil {
+		return nil, err
+	}
+
+	r.modules = mods
+	return mods, err
+}
+
+// DoRelease runs the automated release process, consuming the given Repository's Metadata, updating module's go.mod files,
+// creating a release JSON file, committing changes, tagging the repository, and pushing.
+func (r *Repository) DoRelease(releaseID string) error {
+	if err := r.Metadata.CurrentVersions.isValid(r.git); err != nil {
+		return fmt.Errorf("couldn't create a release: %v", err)
+	}
+
+	enc, bumps, err := r.DiscoverVersions(ReleaseVersionSelector)
+	if err != nil {
+		return fmt.Errorf("couldn't do release: %v", err)
```

Comment:
minor nit: might be useful to have the errors describe a little more about what they were attempting to do before the failure. As an example `failed to discover versions for release: %v`